interface/main/main_screen.php

```diff
@@ -90,87 +89,162 @@ function generate_html_end()
 $errormsg = '';
 
 ///////////////////////////////////////////////////////////////////////
-// Begin code to support U2F.
+// Begin code to support U2F and APP Based TOTP logic.
 ///////////////////////////////////////////////////////////////////////
+function generate_auth_form($requests, $errortype, $errormsg)
+{
+
+    generate_html_start(xl('U2F Key Verification'));
+    echo "<p>\n";
+    echo xlt('Insert your key into a USB port and click the Authenticate button below.');
+    echo " " . xlt('Then press the flashing button on your key within 1 minute.') . "</p>\n";
+    echo "<table><tr><td>\n";
+    echo "<input type='button' value='" . xlt('Authenticate UTF') . "' onclick='doAuth()' />\n";
+    echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "UTF") {
+        echo "<p style='color:red;font-weight:bold'>" . text($errormsg) . "</p>\n";
+    }
+
+    echo '<br /><h3>'.xl('or use').'</h3><br />';
+    echo '<h2>'.xl('TOTP Verification').'</h2>';
+    echo "<p>\n";
+    echo xlt('Enter the code from your authentication application on your device.');
+    echo "<table><tr>";
+    echo "<td><input type='text' name='totp' /></td><td>";
+    echo "<button type='submit'>".xla('Authenticate TOTP')."</button>\n";
+    echo "<input type='hidden' name='form_response' value='true' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "TOTP") {
+        echo "<span style='color:red;font-weight:bold'>" . text($errormsg) . "</span>\n";
+    }
+}
+
 
 $regs = array();          // for mapping device handles to their names
 $registrations = array(); // the array of stored registration objects
 $res1 = sqlStatement(
-    "SELECT a.name, a.var1 FROM login_mfa_registrations AS a " .
-    "WHERE a.user_id = ? AND a.method = 'U2F' ORDER BY a.name",
+    "SELECT a.name, a.method, a.var1 FROM login_mfa_registrations AS a " .
+    "WHERE a.user_id = ? ORDER BY a.name",
     array($_SESSION['authId'])
 );
 while ($row1 = sqlFetchArray($res1)) {
     $regobj = json_decode($row1['var1']);
     $regs[json_encode($regobj->keyHandle)] = $row1['name'];
     $registrations[] = $regobj;
 }
+
 if (!empty($registrations)) {
+    $requests = '';
+    $errortype = '';
     // There is at least one U2F key registered so we have to request or verify key data.
     // https is required, and with a proxy the server might not see it.
     $scheme = "https://"; // isset($_SERVER['HTTPS']) ? "https://" : "http://";
     $appId = $scheme . $_SERVER['HTTP_HOST'];
     $u2f = new u2flib_server\U2F($appId);
```

Comment:
TODO for me:
line above seems odd since don't know if this is for U2F yet

Reply:
The most general case includes multiple U2F keys as well as TOTP; this doesn't hurt anything.
By the way I wonder if there is an intent to support multiple TOTP keys, and if there is any benefit to doing that?

Reply:
Good to know. See no reason to support more than 1 TOTP per users.